src/redisearch_rs/slots_tracker/src/slot_set.rs

```diff
@@ -0,0 +1,781 @@
+    NoMatch, // Set ⊉ query (doesn't cover all)
+}
+
+// ============================================================================
+// Debug assertion helpers
+// ============================================================================
+
+/// Validates that all ranges are valid.
+#[inline]
+fn debug_assert_valid_ranges(ranges: &[SlotRange]) {
+    debug_assert!(ranges.iter().all(|r| r.start <= r.end && r.end <= 16383));
+}
+
+/// Validates that ranges are sorted and normalized (no overlaps or adjacent ranges).
+#[inline]
+fn debug_assert_normalized_ranges(ranges: &[SlotRange]) {
+    debug_assert!(ranges.windows(2).all(|w| w[0].end + 1 < w[1].start));
+}
+
+/// Validates that ranges are valid, sorted, and normalized.
+#[inline]
+fn debug_assert_valid_normalized_input(ranges: &[SlotRange]) {
+    debug_assert_valid_ranges(ranges);
+    debug_assert_normalized_ranges(ranges);
+}
+
+/// A collection of slot ranges with set operation capabilities.
+///
+/// This is an internal type used only within this crate to manage slot sets.
+/// Each range is inclusive [start, end].
+///
+/// **Invariant**: The ranges vector is always kept sorted and normalized:
+/// - Sorted by start slot in ascending order
+/// - No overlapping ranges
+/// - No adjacent ranges (they are merged)
+/// - All ranges are valid (start <= end, values in [0, 16383])
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub(crate) struct SlotSet {
+    /// Vector of slot ranges (start, end), kept sorted and normalized.
+    ranges: Vec<SlotRange>,
+}
+
+impl SlotSet {
+    /// Creates a new empty `SlotSet`.
+    pub(crate) const fn new() -> Self {
+        Self { ranges: Vec::new() }
+    }
+
+    // ========================================================================
+    // Public API methods required for the C FFI:
+    // ========================================================================
+
+    /// Replaces the entire contents of this SlotSet with the given ranges (hard reset).
+    ///
+    /// **Assumes input is already sorted and normalized** (no overlaps, no adjacent ranges).
+    /// In debug builds, validates this assumption.
+    /// Used only by `slots_tracker_set_local_slots`.
+    pub(crate) fn set_from_ranges(&mut self, ranges: &[SlotRange]) {
+        debug_assert_valid_normalized_input(ranges);
+
+        // Input is already normalized, just replace our vector
+        self.ranges = ranges.to_vec();
+    }
+
+    /// Adds/merges ranges into the set (union operation).
+    ///
+    /// **Assumes input is already sorted and normalized**.
+    /// In debug builds, validates this assumption.
+    /// Used by `slots_tracker_set_partially_available_slots` and `slots_tracker_set_fully_available_slots`.
+    pub(crate) fn add_ranges(&mut self, ranges: &[SlotRange]) {
+        debug_assert_valid_normalized_input(ranges);
+
+        // Simply add all ranges and normalize once at the end
+        self.ranges.extend_from_slice(ranges);
+        self.normalize();
```

Comment:
what is the difference between the normalization done in the method w.r.t the normalization asserted at input level?

Reply:
The assert only verifies the input is normalized (in debug mode), while `self.normalize()` normalize `self` if it's not in normal form. Checking if a slot ranges array is normal is simple, while normalizing is a bit more complex.

In this case, we assert the input is valid, and then we append `ranges` (which may invalidate the normal form), and then normalize self again.
There are all kind of edge cases when adding ranges - partial coverage, full coverage, bridging gaps between our ranges and so on. We handle then in `normalize`. We could reuse it in other functionalities in the future.